lib/utils/transaction.js

```diff
@@ -0,0 +1,328 @@
+  // Whenever the v value changes we need to make sure the chainId is also set.
+  const vDescriptors = Object.getOwnPropertyDescriptor(tx, "v");
+  // eslint-disable-next-line accessor-pairs
+  Object.defineProperty(tx, "v", {
+    set: (v) => {
+      vDescriptors.set.call(tx, v);
+      // calculate chainId from signature
+      const sigV = ethUtil.bufferToInt(tx.v);
+      let chainId = Math.floor((sigV - 35) / 2);
+      if (chainId < 0) {
+        chainId = 0;
+      }
+      tx._chainId = chainId || 0;
+    }
+  });
+
+  if (tx.isFake()) {
+    /**
+     * @prop {Buffer} from (read/write) Set from address to bypass transaction
+     * signing on fake transactions.
+     */
+    Object.defineProperty(tx, "from", {
+      enumerable: true,
+      configurable: true,
+      get: tx.getSenderAddress.bind(tx),
+      set: (val) => {
+        if (val) {
+          tx._from = ethUtil.toBuffer(val);
+        } else {
+          tx._from = null;
+        }
+      }
+    });
+
+    if (data && data.from) {
+      tx.from = data.from;
+    }
+
+    tx.hash = fakeHash;
+  }
+}
+
+/**
+ * Parses the given data object and adds its properties to the given tx.
+ * @param {Transaction} tx
+ * @param {Object} [data]
+ */
+function initData(tx, data) {
+  if (data) {
+    if (typeof data === "string") {
+      data = Buffer.from(ethUtil.stripHexPrefix(data), "hex");
+    }
+    if (Buffer.isBuffer(data)) {
+      data = rlp.decode(data);
+    }
+    const self = tx;
+    if (Array.isArray(data)) {
+      if (data.length > tx._fields.length) {
+        throw new Error("wrong number of fields in data");
+      }
+
+      // make sure all the items are buffers
+      data.forEach((d, i) => {
+        self[self._fields[i]] = ethUtil.toBuffer(d);
+      });
+    } else if ((typeof data === "undefined" ? "undefined" : typeof data) === "object") {
+      const keys = Object.keys(data);
+      tx._fields.forEach(function(field) {
+        if (keys.indexOf(field) !== -1) {
+          self[field] = data[field];
+        }
+        if (field === "gasLimit") {
+          if (keys.indexOf("gas") !== -1) {
+            self["gas"] = data["gas"];
+          }
+        } else if (field === "data") {
+          if (keys.indexOf("input") !== -1) {
+            self["input"] = data["input"];
+          }
+        }
+      });
+    } else {
+      throw new Error("invalid data");
+    }
+  }
+}
+
+module.exports = class Transaction extends EthereumJsTransaction {
+  /**
+   * @param {Object} [data] The data for this Transaction.
+   * @param {Number} type The `Transaction.types` bit flag for this transaction
+   *  Can be a combination of `Transaction.types.none`, `Transaction.types.signed`, and `Transaction.types.fake`.
+   */
+  constructor(data, type = Transaction.types.none) {
+    super();
+
+    this.type = type;
+
+    fixProps(this, data);
+    initData(this, data);
+  }
+
+  static get types() {
+    // values must be squares of 2
+    return {
+      none: 0,
```

Comment:
Do we want `none` to be 1  ie: `2^0` instead of 0?

Also I think the comment should read 'powers' of 2.  Ie: `2^n`